### PR TITLE
Tensorly

### LINKED ISSUE #22139: randomized_range_finder

### PR DESCRIPTION
**Pull request by Mr-Niraj-Kulkarni**
close #22139

### CODE REVIEW

File: ivy/data_classes/array/experimental/linear_algebra.py [hunk 1]
```diff
@@ -675,3 +675,29 @@ def tucker(
...
             verbose=verbose,
         )
+
+    def randomized_range_finder(
+        A: Union[ivy.Array, ivy.NativeArray],
+        /,
+        *,
+        n_dims: Optional[int],
+        n_iter: Optional[int] = 2,
+        seed: Optional[int] = None,
```
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/linear_algebra.py:685]: let's add an `out` argument here.

File: ivy/data_classes/array/experimental/linear_algebra.py [hunk 2]
```diff
@@ -675,3 +675,29 @@ def tucker(
...
+    ):
```
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/linear_algebra.py:686]:
```suggestion
    )->ivy.Array:
```

File: ivy/data_classes/array/experimental/linear_algebra.py [hunk 3]
```diff
@@ -675,3 +675,29 @@ def tucker(
...
+        """
+        Compute an orthonormal matrix (Q) whose range approximates the range of A, i.e.,
+        Q Q^H A ≈ A.
+
+        Parameters
+        ----------
+        A : 2D-array
+        n_dims : int, dimension of the returned subspace
```
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/linear_algebra.py:694]:
```suggestion
        n_dims
            dimension of the returned subspace
```
We generally write parameter docstring as:
```
<parameter_name>
       Description of the parameter
```

Comment (hello-fri-end) [reply ivy/data_classes/array/experimental/linear_algebra.py:694]: Would be nice if you could update all the parameters  in the doctstring to follow this format.

File: ivy/data_classes/array/experimental/linear_algebra.py [hunk 4]
```diff
@@ -675,3 +675,29 @@ def tucker(
             tol=tol,
             verbose=verbose,
         )
+
+    def randomized_range_finder(
+        A: Union[ivy.Array, ivy.NativeArray],
```
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/linear_algebra.py:680]:
```suggestion
        self: Union[ivy.Array, ivy.NativeArray],
```

File: ivy/data_classes/array/experimental/linear_algebra.py [hunk 5]
```diff
@@ -675,3 +675,29 @@ def tucker(
...
+        n_dims : int, dimension of the returned subspace
+        n_iter : int, number of power iterations to conduct (default = 2)
+        random_state: {None, int, np.random.RandomState}
+
+        Returns
+        -------
+        Q : 2D-array
+            of shape (A.shape[0], min(n_dims, A.shape[0], A.shape[1]))
+        """
+        return ivy.randomized_range_finder(A, n_dims, n_iter, seed)
```
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/linear_algebra.py:703]:
```suggestion
        return ivy.randomized_range_finder(self, n_dims, n_iter, seed)
```

File: ivy/data_classes/array/experimental/linear_algebra.py [hunk 6]
```diff
@@ -675,3 +675,29 @@ def tucker(
...
+
+        Parameters
+        ----------
+        A : 2D-array
+        n_dims : int, dimension of the returned subspace
+        n_iter : int, number of power iterations to conduct (default = 2)
+        random_state: {None, int, np.random.RandomState}
+
+        Returns
+        -------
```
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/linear_algebra.py:699]: Would you nice if you could add an examples section as well.

File: ivy/data_classes/container/experimental/linear_algebra.py [hunk 7]
```diff
@@ -1889,3 +1889,60 @@ def tucker(
...
         )
+
+    @staticmethod
+    def static_randomized_range_finder(
+        A: Union[ivy.Array, ivy.NativeArray],
+        /,
+        *,
+        n_dims: Optional[int],
+        n_iter: Optional[int] = 2,
+        seed: Optional[int] = None,
```
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/linear_algebra.py:1900]: All of these arguments should have the `ivy.Container` type hint as well. Moreover, there are a few parameters missing in this - please take a look at some functions above for details on missing parameters.

Comment (hello-fri-end) [reply ivy/data_classes/container/experimental/linear_algebra.py:1900]: Also, please add the `out` argument.

File: ivy/data_classes/container/experimental/linear_algebra.py [hunk 8]
```diff
@@ -1889,3 +1889,60 @@ def tucker(
...
+        """
+        Compute an orthonormal matrix (Q) whose range approximates the range of A, i.e.,
+        Q Q^H A ≈ A.
+
+        Parameters
+        ----------
+        A : 2D-array
+        n_dims : int, dimension of the returned subspace
+        n_iter : int, number of power iterations to conduct (default = 2)
+        random_state: {None, int, np.random.RandomState}
```
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/linear_algebra.py:1911]: please update these to follow the format I mentioned.

File: ivy/data_classes/container/experimental/linear_algebra.py [hunk 9]
```diff
@@ -1889,3 +1889,60 @@ def tucker(
...
+
+        Returns
+        -------
+        Q : 2D-array
+            of shape (A.shape[0], min(n_dims, A.shape[0], A.shape[1]))
```
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/linear_algebra.py:1916]: Examples section.

File: ivy/data_classes/container/experimental/linear_algebra.py [hunk 10]
```diff
@@ -1889,3 +1889,60 @@ def tucker(
...
+            "randomized_range_finder", A, n_dims=n_dims, n_iter=n_iter, seed=seed
+        )
+
+    def randomized_range_finder(
+        self: Union[ivy.Array, ivy.NativeArray],
+        /,
+        *,
+        n_dims: Optional[int],
+        n_iter: Optional[int] = 2,
+        seed: Optional[int] = None,
```
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/linear_algebra.py:1928]: Same as above - missing `ivy.Container` type-hint and some parameters are also missing.

File: ivy/data_classes/container/experimental/linear_algebra.py [hunk 11]
```diff
@@ -1889,3 +1889,60 @@ def tucker(
...
+        """
+        Compute an orthonormal matrix (Q) whose range approximates the range of A, i.e.,
+        Q Q^H A ≈ A.
+
+        Parameters
+        ----------
+        A : 2D-array
+        n_dims : int, dimension of the returned subspace
+        n_iter : int, number of power iterations to conduct (default = 2)
+        random_state: {None, int, np.random.RandomState}
```
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/linear_algebra.py:1939]: format update- same as above

File: ivy/data_classes/container/experimental/linear_algebra.py [hunk 12]
```diff
@@ -1889,3 +1889,60 @@ def tucker(
...
+
+        Returns
+        -------
```
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/linear_algebra.py:1942]: Examples section.

File: ivy/functional/backends/jax/experimental/linear_algebra.py [hunk 13]
```diff
@@ -168,7 +168,7 @@ def dot(
     *,
     out: Optional[JaxArray] = None,
 ) -> JaxArray:
-    return jnp.dot(a, b, out=out)
+    return jnp.dot(a, b)
```
Comment (hello-fri-end) [inline ivy/functional/backends/jax/experimental/linear_algebra.py:171]: It seems that there are a lot of problems with `ivy.dot` function. Would  best to create a Trello card for this first and then fix all of these in a separate PR.

File: ivy/functional/frontends/jax/numpy/linalg.py [hunk 14]
```diff
@@ -126,8 +126,3 @@ def tensorinv(a, ind=2):
 @to_ivy_arrays_and_back
 def cond(x, p=None):
     return ivy.cond(x, p=p)
-
-
-@to_ivy_arrays_and_back
-def multi_dot(a):
-    return ivy.multi_dot(a)
```
Comment (hello-fri-end) [inline ivy/functional/frontends/jax/numpy/linalg.py:133]: Why was this deleted?

File: ivy/functional/ivy/experimental/linear_algebra.py [hunk 15]
```diff
@@ -1666,3 +1666,55 @@ def dot(
...
+@inputs_to_ivy_arrays
+@handle_array_function
+@handle_device_shifting
+def randomized_range_finder(
+    A: Union[ivy.Array, ivy.NativeArray],
+    /,
+    *,
+    n_dims: Optional[int],
+    n_iter: Optional[int] = 2,
+    seed: Optional[int] = 0,
```
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/linear_algebra.py:1683]: let's add `out` arguments to the function arguments.

File: ivy_tests/test_ivy/test_frontends/test_jax/test_numpy/test_linalg.py [hunk 16]
```diff
@@ -896,48 +896,41 @@ def test_jax_cond(
...
+    helpers.test_frontend_function(
         input_dtypes=dtype,
         frontend=frontend,
+        backend_to_test=backend_fw,
         test_flags=test_flags,
         fn_tree=fn_tree,
         on_device=on_device,
         test_values=False,
-        args=(x,),
-    )
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_numpy/test_linalg.py:932]: It's not clear to me why this test needs to be modified. If there's a problem with this test then would you please open a separate PR for this? It's not best practice to fix problems with other functions in one PR.

**Review by hello-fri-end — CHANGES_REQUESTED** [review]
Hey @Mr-Niraj-Kulkarni ! Thanks for working on this issue. Have left a few comments - feel free to request another review when you're done with changes.

**Review by hello-fri-end — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_frontends/test_jax/test_numpy/test_linalg.py [hunk 17]
```diff
@@ -892,52 +892,3 @@ def test_jax_cond(
...
-        frontend=frontend,
-        test_flags=test_flags,
-        fn_tree=fn_tree,
-        on_device=on_device,
-        test_values=False,
-        args=(x,),
-    )
-
-    ret = ivy.to_numpy(ret)
-    frontend_ret = np.asarray(frontend_ret)
```
[the remark this answers is not in the document]
Comment (Mr-Niraj-Kulkarni) [reply ivy_tests/test_ivy/test_frontends/test_jax/test_numpy/test_linalg.py:935]: This is from another work I was working on
